pkg/apis/flowcontrol/validation/validation.go

```diff
@@ -0,0 +1,296 @@
+	}
+	for i, rule := range spec.Rules {
+		allErrs = append(allErrs, ValidateFlowSchemaPolicyRuleWithSubjects(&rule, fldPath.Child("rules").Index(i))...)
+	}
+	return allErrs
+}
+
+// ValidateFlowSchemaPolicyRuleWithSubjects validates policy-rule-with-subjects object.
+func ValidateFlowSchemaPolicyRuleWithSubjects(rule *flowcontrol.PolicyRulesWithSubjects, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	if len(rule.Subjects) > 0 {
+		for i, subject := range rule.Subjects {
+			allErrs = append(allErrs, ValidateFlowSchemaSubject(&subject, fldPath.Child("subjects").Index(i))...)
+		}
+	} else {
+		allErrs = append(allErrs, field.Required(fldPath.Child("subjects"), "subjects must contain at least one value"))
+	}
+	if len(rule.ResourceRules) == 0 && len(rule.NonResourceRules) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath, "at least one of resourceRules and nonResourceRules has to be non-empty"))
+	}
+	for i, resourceRule := range rule.ResourceRules {
+		allErrs = append(allErrs, ValidateFlowSchemaResourcePolicyRule(&resourceRule, fldPath.Child("resourceRules").Index(i))...)
+	}
+	for i, nonResourceRule := range rule.NonResourceRules {
+		allErrs = append(allErrs, ValidateFlowSchemaNonResourcePolicyRule(&nonResourceRule, fldPath.Child("nonResourceRules").Index(i))...)
+	}
+	return allErrs
+}
+
+// ValidateFlowSchemaSubject validates flow-schema's subject object.
+func ValidateFlowSchemaSubject(subject *flowcontrol.Subject, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	switch subject.Kind {
+	case flowcontrol.SubjectKindServiceAccount:
+		allErrs = append(allErrs, ValidateServiceAccountSubject(subject.ServiceAccount, fldPath.Child("serviceAccount"))...)
+	case flowcontrol.SubjectKindUser:
+		allErrs = append(allErrs, ValidateUserSubject(subject.User, fldPath.Child("user"))...)
+	case flowcontrol.SubjectKindGroup:
+		allErrs = append(allErrs, ValidateGroupSubject(subject.Group, fldPath.Child("group"))...)
+	default:
+		allErrs = append(allErrs, field.NotSupported(fldPath.Child("kind"), subject.Kind, supportedSubjectKinds.List()))
+	}
+	return allErrs
+}
+
+// ValidateServiceAccountSubject validates subject of "ServiceAccount" kind
+func ValidateServiceAccountSubject(subject *flowcontrol.ServiceAccountSubject, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	if len(subject.Name) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath.Child("name"), ""))
+	}
+	if len(subject.Name) > 0 && subject.Name != flowcontrol.NameAll {
+		for _, msg := range validation.ValidateServiceAccountName(subject.Name, false) {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("name"), subject.Name, msg))
+		}
+	}
+	if len(subject.Namespace) > 0 {
+		for _, msg := range apimachineryvalidation.ValidateNamespaceName(subject.Namespace, false) {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("namespace"), subject.Namespace, msg))
+		}
+	} else {
```

Comment:
Note that for Name we need special treatment for `"*"` but we do not have that for Namespace.

There is a bug here in the check on Name.  If Name is empty the validation should raise a complaint, but doesn't.

Reply:
It could be more uniform: test `len(...) == 0` for both fields, or test `len(...) > 0` for both fields.  Also, I see the error produced in the zero length case is a little different.